### PR TITLE
🚀 Add crafting inventory slots expression (result/matrix)

### PR DESCRIPTION
**Pull request by AyhamAl-Ali**
### Description

🚀 This PR aims to add a new expression for crafting inventory slots such as the `result item` and `matrix`
Some utility methods have been added, if anyone have better way to achieve their results let me know.

**NOTE:** While testing I have found a potential bug when you try to change the `matrix` in `preparing craft` event that will trigger the event again and get into infinite loop, I will try to report that to spigot later when I can but for now I have disabled that behavior in Skript and sent a warning when someone try to use it.

**NOTE:** This PR should be merged after #4409 to be able to use event-inventory

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:** https://github.com/SkriptLang/Skript/issues/4908

### CODE REVIEW

File: src/main/java/ch/njol/skript/util/Utils.java [hunk 1]
```diff
@@ -32,19 +33,15 @@
...
 import org.bukkit.plugin.messaging.Messenger;
 import org.bukkit.plugin.messaging.PluginMessageListener;
 import org.bukkit.util.Vector;
 import org.eclipse.jdt.annotation.Nullable;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Random;
+import java.util.*;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/util/Utils.java:44]: We can't be using wild card in imports. Please disable this feature in your IDE.
This adds imports that we don't need to be importing.

File: src/main/java/ch/njol/skript/aliases/ItemType.java [hunk 2]
```diff
@@ -1288,6 +1288,13 @@ public Material getMaterial() {
 		return data.getType();
 	}
 
+	public ItemStack getItemStack() {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/aliases/ItemType.java:1291]: Why are you adding this method to the ItemType? ItemType has a method called `getRandom()` which generates a random ItemStack based off the ItemType details. The method name is misleading yes, but if we're doing this, this need to deprecate the old method. Please do that in another pull request or use the `getRandom()` method.

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	"on preparing crafting:",
+	"\tset crafting result item to wood"})
+@Since("INSERT VERSION")
+public class ExprCraftingSlots extends SimpleExpression<ItemType> {
+	
+	static {
+		Skript.registerExpression(ExprCraftingSlots.class, ItemType.class, ExpressionType.COMBINED,
+				"[the] crafting [inventory] result (slot|item) [of %inventories%]",
+				"%inventories%'[s] crafting [inventory] result (slot|item)",
+				"[the] crafting [inventory] matrix [(slots|items|shape)] [of %inventories%]",
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:61]: I think we could use another term to describe the shape like crafting grid or just grid.

**Comment by TheLimeGlass** [comment]
I believe we should be using ItemStack for everything. ItemType is used as a literal. This was we can allow for named, lore, enchanted items etc.

This will also remove the need for the Utils `toItemTypes` method, which I personally don't believe we need to be adding.

**Comment by AyhamAl-Ali** [comment]
> I believe we should be using ItemStack for everything. ItemType is used as a literal. This way we can allow for named, lore, enchanted items etc.
> 
> This will also remove the need for the Utils `toItemTypes` method, which I personally don't believe we need to be adding.

This is an important note, I remember I saw a PR about replacing ItemStacks with ItemTypes but didn't follow up with why 
This is the PR #2510 if itemtypes aren't the good choice we better replace them with ItemStacks probably

**Comment by TheLimeGlass** [comment]
It really needs to be explained; So basically ItemTypes are supposed to be used as a Literal. Syntaxes such as `on right click with %itemtypes%` makes sense, because `on rightclick with a diamond sword` can be used. An ItemStack will return all the details of the item, this can be used anywhere in Spigot API, it gives the user more information about the value.

Another example can be in SkQuery recipes, it only accepted ItemTypes, which restricted the user to not allow for named, enchanted or lored items. Where as Tuske later added the ability to use ItemStacks, so you could check if the item was named. This is good for like quests or missions where the item must be named. SkQuery recipes have since been removed.

TPGamesNL also mentioned that we should be doing this for primitive number types. Returning an Integer, or a Double rather than just returning Number. But when it comes to accepting a change, you can be using Number, and just number.intValue() or whatever on it, so the output is certain, and the input can be anything avoiding the cannot understand effect error.

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java (hunk 3 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:61]: I will add `grid` to the pattern but will keep the `matrix` as another option since this is how bukkit named it and just to keep things clear for everyone

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+			} else {
+				items.add(craft.getResult());
+			}
+		}
+		return items.toArray(new ItemStack[0]);
+	}
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:103]: We don't need this switch statement here when we're only checking for 1 mode
```
if (mode == ChangeMode.SET)
    return CollectionUtils.array(ItemStack[].class);
return null;
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	@Override
+	@Nullable
+	protected ItemStack[] get(Event e) {
+		Inventory[] invis = this.invis.getArray(e);
+		if (invis == null)
+			return null;
+
+		List<ItemStack> items = new ArrayList<>(0);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:86]: Small nit picking, but everywhere else in Skript we don't initialize an array list with 0, we just use `new ArrayList<>();` which are the same thing, but for consistency.

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+			}
+		} else { // SET
+			assert invis != null;
+			if (items == null)
+				return;
+			for (Inventory invi : invis) {
+				if (!(invi instanceof CraftingInventory))
+					continue;
+				if (isMatrix) {
+					List<ItemStack> itemStacks = new ArrayList<>(9);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:143]: We could just be using an Array here over an ArrayList as ArrayLists are designed to be expanded based on an unknown amount of values. Here we know it's 9.

This will also improve performance using Arrays over ArrayLists.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java (hunk 4 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java]: Done however, we need 2 modes SET & DELETE

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+				"%inventories%'[s] crafting [inventory] result (slot|item)",
+				"[the] crafting [inventory] (matrix|grid) [(slots|items|shape)] [of %inventories%]",
+				"%inventories%'[s] crafting [inventory] (matrix|grid) [(slots|items|shape)]");
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Inventory> invis;
+	private boolean isMatrix;
+	final ItemStack AIR = new ItemStack(Material.AIR);
+	final ItemStack[] AIR_ITEMSTACK_LIST = new ItemStack[]{AIR, AIR, AIR, AIR, AIR, AIR, AIR, AIR, AIR};
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:67]: These can be private

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+				items.addAll(Arrays.asList(craft.getMatrix()));
+			} else {
+				items.add(craft.getResult());
+			}
+		}
+		return items.toArray(new ItemStack[0]);
+	}
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:100]:
```suggestion
	@Nullable
	public Class<?>[] acceptChange(ChangeMode mode) {
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+		if (mode == ChangeMode.SET || mode == ChangeMode.DELETE)
+			return CollectionUtils.array(ItemStack[].class);
+
+		return null;
+		}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:105]:
```suggestion
	}
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		Inventory[] invis = this.invis.getArray(e);
+		ItemStack[] items = (ItemStack[]) delta;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:110]: You can probably declare this further down (in the else statement)

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+		if (invis == null)
+			return;
+
+		if (e instanceof PrepareItemCraftEvent && isMatrix) { // Until spigot/paper fixes this
+			Skript.warning("Editing the crafting inventory matrix in player " +
+				"preparing craft event will result in an infinite loop due " +
+				"to how spigot/paper handle this event, therefore it's disabled for now.");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:117]:
```suggestion
			Skript.warning("It is not currently possible to edit the crafting inventory matrix in a preparing craft event.");
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+			assert invis != null;
+			if (items == null)
+				return;
+			for (Inventory invi : invis) {
+				if (!(invi instanceof CraftingInventory))
+					continue;
+				if (isMatrix) {
+					ItemStack[] itemStacks = new ItemStack[9];
+					for (int i = 0; i < 9; i++) { // list must be 9 if not we will fill it manually with AIR
+						itemStacks[i] = items.length >= (i+1) ? items[i] : AIR;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:140]:
```suggestion
						itemStacks[i] = i < items.length ? items[i] : AIR;
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.ItemStack;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+@Name("Crafting Inventory Slots")
+@Description({"Represents the slots/items in the crafting inventory such as the result item and matrix of the items."})
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:46]:
```suggestion
@Description("Represents the slots/items in the crafting inventory such as the result item and matrix of the items.")
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+@Examples({"on craft:",
+	"\tif crafting result item is paper:",
+	"\t\tset the crafting matrix to air, air, air, paper, diamond, paper, air, air and air",
+	"on preparing crafting:",
+	"\tset crafting result item to wood"})
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:51]:
```suggestion
@Examples({
	"on craft:",
	"\tif crafting result item is paper:",
	"\t\tset the crafting matrix to air, air, air, paper, diamond, paper, air, air and air",
	"on preparing crafting:",
	"\tset crafting result item to wood"
})
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Nothing super major :D (mainly style suggestions)

I think this will be a useful feature

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+	@Override
+	public Class<? extends ItemStack> getReturnType() {
+		return ItemStack.class;
+	}
+	
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return "crafting " + (isMatrix ? "matrix" : "result") + " slot of " + invis.toString(e, debug);
+	}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:166]:
```suggestion
	}

```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+@Name("Crafting Inventory Slots")
+@Description("Represents the slots/items in the crafting inventory such as the result item and matrix of the items.")
+@Examples({
+	"on craft:",
+	"\tif crafting result item is paper:",
+	"\t\tset the crafting matrix to air, air, air, paper, diamond, paper, air, air and air",
+	"on preparing crafting:",
+	"\tset crafting result item to wood"
+	})
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:53]:
```suggestion
})
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.ItemStack;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+@Name("Crafting Inventory Slots")
+@Description("Represents the slots/items in the crafting inventory such as the result item and matrix of the items.")
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:46]:
```suggestion
@Description("Represents the items in a crafting inventory, such as the items within the crafting grid and the result item.")
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+		}
+		return items.toArray(new ItemStack[0]);
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.DELETE)
+			return CollectionUtils.array(ItemStack[].class);
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:106]:
```suggestion
```

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+	"\tif crafting result item is paper:",
+	"\t\tset the crafting matrix to air, air, air, paper, diamond, paper, air, air and air",
+	"on preparing crafting:",
+	"\tset crafting result item to wood"
+	})
+@Since("INSERT VERSION")
+public class ExprCraftingSlots extends SimpleExpression<ItemStack> {
+	
+	static {
+		Skript.registerExpression(ExprCraftingSlots.class, ItemStack.class, ExpressionType.COMBINED,
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:58]: I think this expression should probably return InventorySlots instead of ItemStacks.

See https://github.com/SkriptLang/Skript/blob/3dbf414e0a16fae3c9ae59d15f0243f2534bf38f/src/main/java/ch/njol/skript/util/slot/InventorySlot.java

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+			ItemStack[] items = (ItemStack[]) delta;
+			if (items == null)
+				return;
+			for (Inventory invi : invis) {
+				if (!(invi instanceof CraftingInventory))
+					continue;
+				if (isMatrix) {
+					ItemStack[] itemStacks = new ItemStack[9];
+					for (int i = 0; i < 9; i++) { // list must be 9 if not we will fill it manually with AIR
+						itemStacks[i] = i < items.length ? items[i] : AIR;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:141]: It's not an issue to use the same itemstack multiple times right?

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+	"\tset crafting result item to wood"
+	})
+@Since("INSERT VERSION")
+public class ExprCraftingSlots extends SimpleExpression<ItemStack> {
+	
+	static {
+		Skript.registerExpression(ExprCraftingSlots.class, ItemStack.class, ExpressionType.COMBINED,
+				"[the] crafting [inventory] result (slot|item) [of %inventories%]",
+				"%inventories%'[s] crafting [inventory] result (slot|item)",
+				"[the] crafting [inventory] (matrix|grid) [(slots|items|shape)] [of %inventories%]",
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:61]: Do you think it would be better to allow the user to access the individual crafting grid slots?
Maybe something like.
`crafting slot %*number% [of %inventories%]`

This may also work better with returning InventorySlots

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java (hunk 20 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java:140]: Shouldn't be

**Comment by sovdeeth** [comment]
Do you still intend to work on this?

**Comment by sovdeeth** [comment]
closing due to inactivity
